### ISSUE TITLE
[Suggestion] Single-file career config

### ISSUE THREAD

**Issue by wrzwicky** [fofix/fofix#103]
I've acquired some new content packs, most of which lack FOF's career info. It would be nice to move all career info to a single file to make it easier to backup up and/or move to updated versions of the pack. It also makes it easy to have alternate careers that refer to songs that might be used in other careers.

Now I don't play career mode, so I'm not sure how (or whether) it works, thus this issue to accumulate discussions how it should work. My initial sketch is below.

I've noticed some games have instrument-specific tiers. I'd recommend these be separate files, but we could include something like `tiers:guitar = gt1 gt2` if we want alternate tiers included in the same file.

```
# key suffixes like ":en-us" :
#   Translated versions of displayable keys to support other languages.
#   Suffix should be an international standard.

[career]
name:en-us = Metallica Career
  # name: friendly display name
subtitle:en = created by me!
  # subtitle: 2nd line under 'name' for author, copyright, etc.
theme = GH Metallica v4
  # theme: Preferred graphics theme while playing this career. Check same
  #   dir as career file first, else regular 'themes' dir.
  # Checking career dir first makes it possible to bundle career, songs, and
  #   theme as one pack.
icon = metallica-load.png       <-- check career dir, else theme dir
  # icon: Check career dir, else theme dir.
format = fofix-career-2
  # format = fofix-career-2 : Must include exactly this string.
tiers = Tier1 Tier2 Tier3
  # tiers: As in titles.ini - list of sections defining career tiers, in
  #   order.
tiers:guitar = gt1 gt2
  # tiers:guitar: Alternate syntax for instrument-specific tiers.

[Tier1]
Name:en = Opening Licks
Unlock_Text:en = You get this one for free.
  # Display if tier is locked.
icon=tier1.png
  # check career, theme, then fofix default dirs
Min_Stars = 0
  # Min_Stars: As in titles.ini.
Song_1 = <hash>
Song_metal = <hash>
  # Hash codes of songs that are part of this tier, in order.
  # <anything> is needed due to limitations of ini files.
  # Scan career dir for songs first. If not found, we'll need to have the
  #   user set the "base library" settings so we can scan all of that.
songs = <hash> <hash>
  # "songs" keys can list several.
  # All song/songs keys are processed strictly in file order to produce the
  #   default ordering. The _anything is ignored.
  # No need to be picky about parsing; Both song and songs keys can contain
  #   zero, one, or many hashes.  Any white space or punctuation can be used
  #   to separate them.

[Tier2]
Name:en = Axe-Grinders
icon=tier2.png
Unlock_Require = Tier1
Unlock_Text:en = Beat "Opening Licks" with 10 stars to unlock.
Min_Stars = 10
Songs=<hash> <hash>
Song_etc = etc
songs_etc = etc etc
```

**Comment by mdsitton** [comment]
Yeah, career mode is not something I typically play on FoFiX (or any rhythm game tbh) so it basically never gets tested. I've never even looked into how career's work...

You do also bring up some points here that have been bugging me for a while about fofix, which is basically that the rockmeter.ini, Song code, and the custom theme code doesn't have any clearly defined localization support nor any guides and documentation about how one would go about supporting it in a theme. Which i think should be a separate issue most likely.

Back to the above format.

If ini files are limiting features that could be implemented for this, i'd suggest we use something else such as json or yaml?

**Comment by wrzwicky** [comment]
My first sketch was XML, but INI actually works quite nicely for this.  Actually using these files would require much work, but it might be worth trying the :language thing on other file first.

**Comment by mdsitton** [comment]
I really really dislike xml by the way. But if anything i'd rather use a data format here that doesn't require a lot of manual string manipulation on our end when implementing it.

**Comment by Linkid** [comment]
I like your idea about language. And I think using ini file is enough for what we do (and others use this format, so we should keep compatibility). However, after a quick check, using two delimiters (`:` and `=`) is not possible as is with the standard lib…

**Comment by nhydock** [comment]
As the project is written in python, I'd prefer in terms of configuration files, it'd be nice to start making new configuration in JSON since it's much better supported and widely used.  Especially since you can easily do nesting and arrays with the format.  It was not an option at the time FoF was made since the json module was not added until Python 2.6, but with new development there's little reason not to use it for stuff like this.

https://docs.python.org/2/library/json.html

**Comment by wrzwicky** [comment]
@nhydock Good idea, I avoided JSON because we're already using ini files, plus JSON does not support comments. But for discussion, attached below is a JSON file example. Changes from above:
- "name" field is now called "label". I don't like using "name" for things that cannot be canonical.
- Localization is a dictionary inside the field.  I'm not sure I like this; it might be easier for users to cope with the "label:en-us" style syntax.
- JSON does not support comments, so we need to weave comment-like info into the file format, and update the parser accordingly. For example, song hash strings contain the song name.

[_generated_career_.json.txt](https://github.com/fofix/fofix/files/1412610/_generated_career_.json.txt)

FYI: I made a tool to generate these career files, it's in my repo: 
 d001ee35b2437be40cc607f9497ce6d12aaa61f2

**Comment by nhydock** [comment]
why, exactly, are comments necessary in a data format?  It's fine if we have the spec documented somewhere, but the data itself doesn't need comments.

**Comment by Linkid** [comment]
Why do you change the file format ? Could you summary your idea in a list, please ? I don't really see the point here. The only thing I understood is translating names.

Furthermore, I explicitly removed the wiki page to put all the doc in the source. Activating it again will be confusing.

**Comment by nhydock** [comment]
If there's going to be breaking changes with the project, then there's no point in looking compatible.  Might as well move to something that's better supported, more flexible, and easier to read when defining relational data.  It's been listed as a todo for the project for a long while to start moving away from ini files in a lot of places.  They only exist as a hold over from FoF, not because they're nice to work with.
